### PR TITLE
Module collision check on generated item only

### PR DESCRIPTION
**Pull request by chunlokling-da**
fixes [#15340](https://github.com/digital-asset/daml/issues/15340)

```rst
CHANGELOG_BEGIN
Bug fix: Module name collision check performed on modules which is not to be generated https://github.com/digital-asset/daml/issues/15340 https://github.com/digital-asset/daml/issues/15341
CHANGELOG_END
```

### CODE REVIEW

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala [hunk 1]
```diff
@@ -23,6 +23,9 @@ final case class TransitiveClosure(
 object TransitiveClosure {
 
   def from(orderedDependencies: OrderedDependencies[Identifier, NodeType]): TransitiveClosure = {
+    orderedDependencies.deps.foreach { case (id, node) =>
+      println(s"$id -> $node")
+    }
```
Comment (chunlokling-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala:28]: debug log should be deleted

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 2]
```diff
@@ -285,4 +285,5 @@ object Ref {
   type TypeSynName = Identifier
   val TypeSynName = Identifier
 
+  case class ModuleId(packageId: PackageId, moduleName: ModuleName)
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:288]: I am not very queen on adding this globally for a very limited purpose. 
Please define more locally where you need or use daml.lf.language.Reference.Module

Comment (fayi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:288]: @remyhaemmerle-da this is just temporary.

@chunlokling-da let's mark the PR as draft for now

Comment (fayi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:288]: Good shout on `daml.lf.language.Reference.Module`. We can use that

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala (hunk 1 above)
Comment (fayi-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala]: Removed

File: language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala [hunk 3]
```diff
@@ -42,6 +42,20 @@ final class CodeGenRunnerTests extends AnyFlatSpec with Matchers {
     assert(scope.toBeGenerated === Set.empty)
   }
 
+  // Test case reproducing #15341
+  it should "read interfaces from 2 DAR files with same dependencies but one with different daml compiler version" in {
```
Comment (chunlokling-da) [inline language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala:46]: // Test case reproducing #15341

**Review by ray-roestenburg-da — APPROVED** [review]

File: language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala [hunk 4]
```diff
@@ -54,38 +68,75 @@ final class CodeGenRunnerTests extends AnyFlatSpec with Matchers {
...
+    )
+  }
+
+  it should "succeed if same module name between a module not to be generated and a module to be generated " in {
+    val signatures = Seq(interface("pkg1", "A"), interface("pkg2", "A"))
+    assert(
+      CodeGenRunner.detectModuleCollisions(
+        Map.empty,
+        signatures,
+        Set(Reference.Module(PackageId.assertFromString("pkg1"), ModuleName.assertFromString("A"))),
```
Comment (chunlokling-da) [inline language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala:139]: new test cases to test the behavior of module name collision when generated module information is provided

**Pull request merged**